### PR TITLE
fix: det tunnel should work with proxy port exposed [FE-121]

### PR DESCRIPTION
**Pull request by rcorujo**
## Description

**Note:** Some e2e tests, such as `tests/command/test_run.py::test_image_pull_after_remove` use an image (e.g., `environment.image=archlinux:base-20211010.0.36274`) that does not contain Python.  Therefore, in order to prevent the test from failing, I've changed `command-entrypoint.sh` to not exit if Python is not installed.

Anoop reported that when running a command from the MAC laptop with the Determined devcluster (i.e., `tools/slurmcluster.sh`) pointing to the `grenoble` cluster, tunneling wasn't working when the proxy port was configured.

To reproduce the issue observed by Anoop, perform the following steps:

1. On your MAC laptop, start the dev launcher and point it to the `grenoble` cluster.

```
(base) rcorujo@C02FJ0T8MD6Q hpc-ard-capsules-core % ./loadDevLauncher.sh -d o184i054.gre.smktg.hpecorp.net
```

2. On your MAC laptop, start the Determined devcluster and point it to the `grenoble` cluster.

```
(determined) (base) rcorujo@C02FJ0T8MD6Q determined-ee % tools/slurmcluster.sh -d o184i054
```

3. Create the following `config-client.yaml` file. It references the `mlde_cpus` pool on the `grenoble` cluster.  If you are pointing to a different cluster, then use a resource pool that is appropriate for that cluster.  For example, if you were referencing the `casablanca-login` cluster, you would use the resource pool `cpuQ`.

```
resources:
  resource_pool: mlde_cpus
environment:
  environment_variables:
    - https_proxy=http://web-proxy.bbn.hpecorp.net:8088
    - http_proxy=http://web-proxy.bbn.hpecorp.net:8088
    - GRADIO_NUM_PORTS=1
    - MODEL_INF_END_POINT=http://16.16.186.225:8080
  proxy_ports:
    - proxy_port: 7860
      proxy_tcp: true
      unauthenticated: true
  image:
    cpu: anoophpe/tgi-client:1.0 
```

4. Run the command using the `config-client.yaml` file shown above.

```
det command run --config-file config-client.yaml "cd /app && python app.py"
Launched command (id: 06e05fd4-c482-48d5-9dda-6c37f81543ec).
...
```

5. Start your Determined tunnel, which is supposed to allow you to access a port on the container by referencing `localhost:port` from your MAC laptop.  Make sure you set your `DET_MASTER` environment variable appropriately (e.g.,`DET_MASTER=localhost:8081`) and set `TASK_ID` to the ID returned by the command (e.g., `06e05fd4-c482-48d5-9dda-6c37f81543ec` in the example above).

```
python -m determined.cli.tunnel --listener 7860 $DET_MASTER $TASK_ID:7860
```
6. Trying to access `http://localhost:7860` from a browser on your MAC laptop will fail.

7. Similarly, running `wget http://localhost:7860` from a bash shell on your MAC laptop will also fail.

The root cause of problem is two-fold. 

The first issue is that the `command-entrypoint.sh` does not contain the `prep_container --proxy`, like the `shell-entrypoint.sh` does, so the proxies are not registered in the master.

However, even after adding `prep_container --proxy` to `command-entrypoint.sh`, an attempt to connect to `http://localhost:7860` still fails.

To further analyze the second issue as to why we still can't connect, we have to exec into Anoop's container to see what the `gradio` application that we're trying to connect to is doing.  We can do this from our MAC laptop.

```
docker run -it anoophpe/tgi-client:1.0 /bin/bash 
```

In `/opt/conda/lib/python3.8/site-packages/gradio/networking.py` inside the container, we can see that it’s binding to `127.0.0.1` if the `GRADIO_SERVER_NAME` environment variable is not set.  This means that the `gradio` application will only accept connections originating from the localhost, which explains the reason why we cannot connect.

```
INITIAL_PORT_VALUE = int(os.getenv("GRADIO_SERVER_PORT", "7860"))
TRY_NUM_PORTS = int(os.getenv("GRADIO_NUM_PORTS", "100"))
LOCALHOST_NAME = os.getenv("GRADIO_SERVER_NAME", "127.0.0.1")

...

         s = socket.socket()
         s.bind((LOCALHOST_NAME, server_port))
```

To resolve this, we can set `GRADIO_SERVER_NAME=0.0.0.0` in the `config-client.yaml` to bind to all available IPv4 addresses (i.e., accept connections from any network interface).

```
resources:
  resource_pool: mlde_cpus
  #resource_pool: cpuQ
environment:
  environment_variables:
    - https_proxy=http://web-proxy.bbn.hpecorp.net:8088
    - http_proxy=http://web-proxy.bbn.hpecorp.net:8088
    - GRADIO_NUM_PORTS=1
    - MODEL_INF_END_POINT=http://16.16.186.225:8080
    - GRADIO_SERVER_NAME=0.0.0.0
  proxy_ports:
    - proxy_port: 7860
      proxy_tcp: true
      unauthenticated: true
  image:
    cpu: anoophpe/tgi-client:1.0 
```

Now when I run the Determined command, you can see it says `Running on local URL:  http://0.0.0.0:7860`.

```
(base) rcorujo@C02FJ0T8MD6Q documents % PYTHONPATH=$HOME/IdeaProjects/FE-121_det_tunnel_should_work_with_proxy_port_exposed/determined-ee/harness  DET_MASTER=localhost:8081 det command run --config-file config-client_rig.yaml "cd /app && python app.py"
Launched command (id: b7d90efd-044d-4fde-b180-b9e615701cd9).
[2023-07-26T21:32:09.962151Z]          || INFO: Scheduling Command (virtually-rare-collie) (id: b7d90efd-044d-4fde-b180-b9e615701cd9)
[2023-07-26T21:32:20.797635Z]          || INFO: HPC Job ID: 138
[2023-07-26T21:32:20.797766Z]          || INFO: Command (virtually-rare-collie) was assigned to an agent
[2023-07-26T21:32:25.055339Z] [cont=0] || WARNING: [302891] root: falling back to naive proxy ip resolution (error=None)
[2023-07-26T21:32:30.661529Z]          || INFO: Resources for Command (virtually-rare-collie) have started
[2023-07-26T21:32:34.007477Z] [cont=0] || Running on local URL:  http://0.0.0.0:7860
[2023-07-26T21:32:34.007575Z] [cont=0] || 
[2023-07-26T21:32:34.007612Z] [cont=0] || To create a public link, set `share=True` in `launch()`.
```

Then I start the Determined tunnel.

```
(determined) (base) rcorujo@C02FJ0T8MD6Q NOTES % python3 -m determined.cli.tunnel --listener 7860 $DET_MASTER b7d90efd-044d-4fde-b180-b9e615701cd9:7860
```

From my MAC laptop, I am now able to connect to the `gradio` application running on the compute node.

```
(base) rcorujo@C02FJ0T8MD6Q ~ % wget http://localhost:7860
--2023-07-26 17:33:14--  http://localhost:7860/
Resolving localhost (localhost)... ::1, 127.0.0.1
Connecting to localhost (localhost)|::1|:7860... failed: Connection refused.
Connecting to localhost (localhost)|127.0.0.1|:7860... connected.
HTTP request sent, awaiting response... 200 OK
Length: 7428 (7.3K) [text/html]
Saving to: ‘index.html’

index.html                                                 100%[========================================================================================================================================>]   7.25K  --.-KB/s    in 0s      

2023-07-26 17:33:14 (354 MB/s) - ‘index.html’ saved [7428/7428]
```

I can also now connect to the UI from a browser on my MAC laptop.

image

## Test Plan

Performed the steps mentioned above on both the `casablanca-login` and `grenoble` clusters.  In both cases, I was able to connect to the `gradio` application that's running in Anoop's container on the compute node using`http://localhost:7860` from a browser on my MAC laptop.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

**Comment by jerryharrow** [comment]
Looks good, but the test failures are concerning.    My e2e tests from this morning looked good.

One failure is actually of test_logging.py which uses a command.   Should get the tests cleaner before approval.

https://app.circleci.com/pipelines/github/determined-ai/determined/42562/workflows/18e046c1-f9a4-49e9-b2ef-5c4783071224/jobs/1674612

**Comment by jerryharrow** [comment]
main looks like it should be pretty much green from the e2etest **https://app.circleci.com/pipelines/github/determined-ai/determined?branch=main
**

File: master/static/srv/command-entrypoint.sh
Comment (phillip-gaisford) [inline master/static/srv/command-entrypoint.sh:6]: We can't rely on python being available in the container?

Comment (rcorujo) [reply master/static/srv/command-entrypoint.sh:6]: Just copied what the other entrypoint.sh scripts were doing.

Comment (phillip-gaisford) [reply master/static/srv/command-entrypoint.sh:6]: Seems like a mess -- who exactly would `install python3` at this point?

Comment (rcorujo) [reply master/static/srv/command-entrypoint.sh:6]: Not sure.  Maybe you can't install it on the host and set the DET_PYTHON_EXECUTABLE to the path where you've bind mounted the directory on the container?

**Comment by rcorujo** [comment]
> Looks good, but the test failures are concerning. My e2e tests from this morning looked good.
> 
> One failure is actually of test_logging.py which uses a command. Should get the tests cleaner before approval.
> 
> https://app.circleci.com/pipelines/github/determined-ai/determined/42562/workflows/18e046c1-f9a4-49e9-b2ef-5c4783071224/jobs/1674612

Submitted another test PR that only added a single comment line and it's still failing the same way, so I don't think that the failures have anything to do with my code change.

https://github.com/determined-ai/determined/pull/7495

**Comment by rcorujo** [comment]
> > Looks good, but the test failures are concerning. My e2e tests from this morning looked good.
> > One failure is actually of test_logging.py which uses a command. Should get the tests cleaner before approval.
> > https://app.circleci.com/pipelines/github/determined-ai/determined/42562/workflows/18e046c1-f9a4-49e9-b2ef-5c4783071224/jobs/1674612
> 
> Submitted another test PR that only added a single comment line and it's still failing the same way, so I don't think that the failures have anything to do with my code change.
> 
> #7495

This person's PR also has the same failures.

https://github.com/determined-ai/determined/pull/7491

**Comment by rcorujo** [comment]
> > > Looks good, but the test failures are concerning. My e2e tests from this morning looked good.
> > > One failure is actually of test_logging.py which uses a command. Should get the tests cleaner before approval.
> > > https://app.circleci.com/pipelines/github/determined-ai/determined/42562/workflows/18e046c1-f9a4-49e9-b2ef-5c4783071224/jobs/1674612
> > 
> > 
> > Submitted another test PR that only added a single comment line and it's still failing the same way, so I don't think that the failures have anything to do with my code change.
> > #7495
> 
> This person's PR also has the same failures.
> 
> #7491

Seems to have been reported on Slack:

https://hpe-aiatscale.slack.com/archives/C04C9JXB1C2/p1690325131819849

File: master/static/srv/command-entrypoint.sh
Comment (jerryharrow) [reply master/static/srv/command-entrypoint.sh:6]: People can build their own environment files -- so they might install it as python vs python3.   There is no standardization here AFAICT. :-( 


So make sense to be consistent with other entrypoint scripts.

**Review by phillip-gaisford — APPROVED** [review]
Excellent write-up!

**Comment by jerryharrow** [comment]
I would rebase again.   main tests are mostly clean as compared to this branch.

https://app.circleci.com/pipelines/github/determined-ai/determined?branch=main

File: master/static/srv/command-entrypoint.sh
Comment (stoksc) [reply master/static/srv/command-entrypoint.sh:6]: the log shipper already relies on python and already runs this logic, so we are fine.

File: master/static/srv/command-entrypoint.sh
Comment (stoksc) [inline master/static/srv/command-entrypoint.sh:17]: doesn't have to be in this ticket, but the trial entrypoint, `entrypoint.sh` in this folder, suffers the same bug

Comment (rcorujo) [reply master/static/srv/command-entrypoint.sh:17]: I had originally made a similar change in "entrypoint.sh", but when I ran an experiment, I got an error, so I pulled out the change.

```
<info> [2023-07-27 15:37:12] || INFO: Trial 152990 (Experiment 11277) was assigned to an agent
<info> [2023-07-27 15:37:20] [cont=0] || INFO: [1701104] root: detected 1 gpus
<info> [2023-07-27 15:37:20] [cont=0] || INFO: [1701104] root: Running task container on agent_id=node002, hostname=node002 with visible GPUs ['GPU-ee310a06-c7fb-f921-f137-d605913909b1']
<warning> [2023-07-27 15:37:21] [cont=0] || WARNING: [1701104] root: falling back to naive proxy ip resolution (error=None)
<none> [2023-07-27 15:37:22] [cont=0] || Traceback (most recent call last):
<none> [2023-07-27 15:37:22] [cont=0] ||   File "/opt/conda/lib/python3.8/runpy.py", line 194, in _run_module_as_main
<none> [2023-07-27 15:37:22] [cont=0] ||     return _run_code(code, main_globals, None,
<none> [2023-07-27 15:37:22] [cont=0] ||   File "/opt/conda/lib/python3.8/runpy.py", line 87, in _run_code
<none> [2023-07-27 15:37:22] [cont=0] ||     exec(code, run_globals)
<none> [2023-07-27 15:37:22] [cont=0] ||   File "/run/determined/pythonuserbase/lib/python3.8/site-packages/determined/exec/prep_container.py", line 358, in <module>
<none> [2023-07-27 15:37:22] [cont=0] ||     do_proxy(sess, info.allocation_id)
<none> [2023-07-27 15:37:22] [cont=0] ||   File "/run/determined/pythonuserbase/lib/python3.8/site-packages/determined/exec/prep_container.py", line 281, in do_proxy
<none> [2023-07-27 15:37:22] [cont=0] ||     set_proxy_address(sess, allocation_id)
<none> [2023-07-27 15:37:22] [cont=0] ||   File "/run/determined/pythonuserbase/lib/python3.8/site-packages/determined/exec/prep_container.py", line 265, in set_proxy_address
<none> [2023-07-27 15:37:22] [cont=0] ||     bindings.post_PostAllocationProxyAddress(
<none> [2023-07-27 15:37:22] [cont=0] ||   File "/run/determined/pythonuserbase/lib/python3.8/site-packages/determined/common/api/bindings.py", line 16610, in post_PostAllocationProxyAddress
<none> [2023-07-27 15:37:22] [cont=0] ||     _resp = session._do_request(
<none> [2023-07-27 15:37:22] [cont=0] ||   File "/run/determined/pythonuserbase/lib/python3.8/site-packages/determined/common/api/_session.py", line 36, in _do_request
<none> [2023-07-27 15:37:22] [cont=0] ||     return request.do_request(
<none> [2023-07-27 15:37:22] [cont=0] ||   File "/run/determined/pythonuserbase/lib/python3.8/site-packages/determined/common/api/request.py", line 180, in do_request
<none> [2023-07-27 15:37:22] [cont=0] ||     raise errors.APIException(r)
<none> [2023-07-27 15:37:22] [cont=0] || determined.common.api.errors.APIException: {"error":{"code":13,"reason":"Internal","error":"task.SetAllocationProxyAddress not supported for this allocation or resource manager"}}
<none> [2023-07-27 15:37:22] [cont=0] ||
<info> [2023-07-27 15:37:22] || INFO: Resources for Trial 152990 (Experiment 11277) have started
<error> [2023-07-27 15:38:34] || ERROR: Job failed in state FAILED. The job terminated with a non-zero exit code.
Job terminated. Most recent output from the stderr log is: INFO:    Using cached SIF image
WARNING: Overriding HOME environment variable with APPTAINERENV_HOME is not permitted INFO: Setting workdir to /run/determined/workdir
INFO: executing /run/determined/train/entrypoint.sh INFO: [1701104] root: detected 1 gpus
INFO: [1701104] root: Running task container on agent_id=node002, hostname=node002 with visible GPUs ['GPU-ee310a06-c7fb-f921-f137-d605913909b1'] WARNING: [1701104] root: falling back to naive proxy ip resolution (error=None)
Traceback (most recent call last):   File "/opt/conda/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,   File "/opt/conda/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)   File "/run/determined/pythonuserbase/lib/python3.8/site-packages/determined/exec/prep_container.py", line 358, in <module>
    do_proxy(sess, info.allocation_id)   File "/run/determined/pythonuserbase/lib/python3.8/site-packages/determined/exec/prep_container.py", line 281, in do_proxy
    set_proxy_address(sess, allocation_id)   File "/run/determined/pythonuserbase/lib/python3.8/site-packages/determined/exec/prep_container.py", line 265, in set_proxy_address
    bindings.post_PostAllocationProxyAddress(   File "/run/determined/pythonuserbase/lib/python3.8/site-packages/determined/common/api/bindings.py", line 16610, in post_PostAllocationProxyAddress
    _resp = session._do_request(   File "/run/determined/pythonuserbase/lib/python3.8/site-packages/determined/common/api/_session.py", line 36, in _do_request
    return request.do_request(   File "/run/determined/pythonuserbase/lib/python3.8/site-packages/determined/common/api/request.py", line 180, in do_request
    raise errors.APIException(r) determined.common.api.errors.APIException: {"error":{"code":13,"reason":"Internal","error":"task.SetAllocationProxyAddress not supported for this allocation or resource manager"}}
 srun: error: node002: task 0: Exited with exit code 1
srun: Terminating StepId=174.0 
```

Comment (stoksc) [reply master/static/srv/command-entrypoint.sh:21]: thanks for the stack. i'll file a ticket for my team.

**Review by jerryharrow — APPROVED** [review]

**Pull request merged**
